File: backend/api/schemas/scheduled_report.py
"""Pydantic schemas for scheduled report API endpoints.

NEM-3621: Create Scheduled Reports API Schemas and Routes

Provides schemas for:
- Creating scheduled reports (daily, weekly, monthly)
- Updating scheduled report configurations
- Response schemas with report metadata
- List response with pagination
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

__all__ = [
    "ReportFormat",
    "ReportFrequency",
    "ScheduledReportCreate",
    "ScheduledReportListResponse",
    "ScheduledReportResponse",
    "ScheduledReportRunResponse",
    "ScheduledReportUpdate",
]


class ReportFrequency(str, Enum):
    """Frequency options for scheduled reports.

    Values:
        DAILY: Report runs every day at specified time
        WEEKLY: Report runs on specified day of week
        MONTHLY: Report runs on specified day of month
    """

    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"

    def __str__(self) -> str:
        """Return string representation of frequency."""
        return self.value


class ReportFormat(str, Enum):
    """Output format options for scheduled reports.

    Values:
        PDF: Portable Document Format with charts and summaries
        CSV: Comma-separated values for data analysis
        JSON: Machine-readable JSON format
    """

    PDF = "pdf"
    CSV = "csv"
    JSON = "json"

    def __str__(self) -> str:
        """Return string representation of format."""
        return self.value


class ScheduledReportCreate(BaseModel):
    """Schema for creating a new scheduled report.

    Used when setting up a new automated report schedule.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "name": "Weekly Security Summary",
                "frequency": "weekly",
                "day_of_week": 1,
                "hour": 8,
                "minute": 0,
                "timezone": "America/New_York",
                "format": "pdf",
                "enabled": True,
                "email_recipients": ["admin@example.com"],
                "include_charts": True,
                "include_event_details": True,
            }
        }
    )

    name: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Name/title of the scheduled report",
    )
    frequency: ReportFrequency = Field(
        ...,
        description="How often the report should run (daily, weekly, monthly)",
    )
    day_of_week: int | None = Field(
        None,
        ge=0,
        le=6,
        description="Day of week (0=Monday, 6=Sunday) for weekly reports",
    )
    day_of_month: int | None = Field(
        None,
        ge=1,
        le=31,
        description="Day of month (1-31) for monthly reports",
    )
    hour: int = Field(
        8,
        ge=0,
        le=23,
        description="Hour of day to run report (0-23, default 8)",
    )
    minute: int = Field(
        0,
        ge=0,
        le=59,
        description="Minute of hour to run report (0-59, default 0)",
    )
    timezone: str = Field(
        "UTC",
        max_length=50,
        description="Timezone for schedule (e.g., 'America/New_York', 'UTC')",
    )
    format: ReportFormat = Field(
        ReportFormat.PDF,
        description="Output format for the report (pdf, csv, json)",
    )
    enabled: bool = Field(
        True,
        description="Whether the scheduled report is active",
    )
    email_recipients: list[str] | None = Field(
        None,
        max_length=10,
        description="Email addresses to send report to (max 10)",
    )
    include_charts: bool = Field(
        True,
        description="Include visual charts in the report",
    )
    include_event_details: bool = Field(
        True,
        description="Include detailed event breakdowns",
    )

    @field_validator("email_recipients")
    @classmethod
    def validate_email_recipients(cls, v: list[str] | None) -> list[str] | None:
        """Validate email recipients are properly formatted."""
        if v is None:
            return v
        if len(v) > 10:
            raise ValueError("Maximum of 10 email recipients allowed")
        for email in v:
            if "@" not in email or len(email) > 254:
                raise ValueError(f"Invalid email format: {email}")
        return v

    @model_validator(mode="after")
    def validate_schedule_fields(self) -> ScheduledReportCreate:
        """Validate that schedule fields match the frequency.

        - Weekly reports require day_of_week
        - Monthly reports require day_of_month
        - Daily reports should not have day_of_week or day_of_month
        """
        if self.frequency == ReportFrequency.WEEKLY:
            if self.day_of_week is None:
                raise ValueError("day_of_week is required for weekly reports")
        elif self.frequency == ReportFrequency.MONTHLY:
            if self.day_of_month is None:
                raise ValueError("day_of_month is required for monthly reports")

        return self


class ScheduledReportUpdate(BaseModel):
    """Schema for updating an existing scheduled report.

    All fields are optional for partial updates.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "name": "Updated Weekly Report",
                "enabled": False,
            }
        }
    )

    name: str | None = Field(
        None,
        min_length=1,
        max_length=255,
        description="Name/title of the scheduled report",
    )
    frequency: ReportFrequency | None = Field(
        None,
        description="How often the report should run (daily, weekly, monthly)",
    )
    day_of_week: int | None = Field(
        None,
        ge=0,
        le=6,
        description="Day of week (0=Monday, 6=Sunday) for weekly reports",
    )
    day_of_month: int | None = Field(
        None,
        ge=1,
        le=31,
        description="Day of month (1-31) for monthly reports",
    )
    hour: int | None = Field(
        None,
        ge=0,
        le=23,
        description="Hour of day to run report (0-23)",
    )
    minute: int | None = Field(
        None,
        ge=0,
        le=59,
        description="Minute of hour to run report (0-59)",
    )
    timezone: str | None = Field(
        None,
        max_length=50,
        description="Timezone for schedule (e.g., 'America/New_York', 'UTC')",
    )
    format: ReportFormat | None = Field(
        None,
        description="Output format for the report (pdf, csv, json)",
    )
    enabled: bool | None = Field(
        None,
        description="Whether the scheduled report is active",
    )
    email_recipients: list[str] | None = Field(
        None,
        max_length=10,
        description="Email addresses to send report to (max 10)",
    )
    include_charts: bool | None = Field(
        None,
        description="Include visual charts in the report",
    )
    include_event_details: bool | None = Field(
        None,
        description="Include detailed event breakdowns",
    )

    @field_validator("email_recipients")
    @classmethod
    def validate_email_recipients(cls, v: list[str] | None) -> list[str] | None:
        """Validate email recipients are properly formatted."""
        if v is None:
            return v
        if len(v) > 10:
            raise ValueError("Maximum of 10 email recipients allowed")
        for email in v:
            if "@" not in email or len(email) > 254:
                raise ValueError(f"Invalid email format: {email}")
        return v


class ScheduledReportResponse(BaseModel):
    """Schema for scheduled report response.

    Returned when retrieving or creating a scheduled report.
    """

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": 1,
                "name": "Weekly Security Summary",
                "frequency": "weekly",
                "day_of_week": 1,
                "day_of_month": None,
                "hour": 8,
                "minute": 0,
                "timezone": "America/New_York",
                "format": "pdf",
                "enabled": True,
                "email_recipients": ["admin@example.com"],
                "include_charts": True,
                "include_event_details": True,
                "last_run_at": "2025-01-20T08:00:00Z",
                "next_run_at": "2025-01-27T08:00:00Z",
                "created_at": "2025-01-01T12:00:00Z",
                "updated_at": "2025-01-15T09:30:00Z",
            }
        },
    )

    id: int = Field(..., description="Scheduled report ID")
    name: str = Field(..., description="Name/title of the scheduled report")
    frequency: ReportFrequency | str = Field(..., description="How often the report runs")
    day_of_week: int | None = Field(
        None, description="Day of week (0=Monday, 6=Sunday) for weekly reports"
    )
    day_of_month: int | None = Field(None, description="Day of month (1-31) for monthly reports")
    hour: int = Field(..., description="Hour of day to run report (0-23)")
    minute: int = Field(..., description="Minute of hour to run report (0-59)")
    timezone: str = Field(..., description="Timezone for schedule")
    format: ReportFormat | str = Field(..., description="Output format for the report")
    enabled: bool = Field(..., description="Whether the scheduled report is active")
    email_recipients: list[str] | None = Field(
        None, description="Email addresses to send report to"
    )
    include_charts: bool = Field(..., description="Include visual charts in the report")
    include_event_details: bool = Field(..., description="Include detailed event breakdowns")
    last_run_at: datetime | None = Field(None, description="When the report last ran successfully")
    next_run_at: datetime | None = Field(
        None, description="When the report is scheduled to run next"
    )
    created_at: datetime = Field(..., description="When the report was created")
    updated_at: datetime = Field(..., description="When the report was last updated")


class ScheduledReportListResponse(BaseModel):
    """Schema for scheduled report list response.

    Returns a list of scheduled reports with total count.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "items": [
                    {
                        "id": 1,
                        "name": "Weekly Security Summary",
                        "frequency": "weekly",
                        "day_of_week": 1,
                        "day_of_month": None,
                        "hour": 8,
                        "minute": 0,
                        "timezone": "America/New_York",
                        "format": "pdf",
                        "enabled": True,
                        "email_recipients": ["admin@example.com"],
                        "include_charts": True,
                        "include_event_details": True,
                        "last_run_at": "2025-01-20T08:00:00Z",
                        "next_run_at": "2025-01-27T08:00:00Z",
                        "created_at": "2025-01-01T12:00:00Z",
                        "updated_at": "2025-01-15T09:30:00Z",
                    }
                ],
                "total": 1,
            }
        }
    )

    items: list[ScheduledReportResponse] = Field(..., description="List of scheduled reports")
    total: int = Field(..., ge=0, description="Total number of scheduled reports")


class ScheduledReportRunResponse(BaseModel):
    """Schema for manual report run response.

    Returned when a report is manually triggered.
    """

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "report_id": 1,
                "status": "running",
                "message": "Report generation started",
                "started_at": "2025-01-25T10:30:00Z",
            }
        }
    )

    report_id: int = Field(..., description="ID of the report being run")
    status: str = Field(..., description="Status of the run (running, queued, failed)")
    message: str = Field(..., description="Status message")
    started_at: datetime = Field(..., description="When the run was initiated")

File: frontend/src/types/scheduledReport.ts
/**
 * Scheduled Reports Types
 *
 * Types for scheduled report management feature.
 * These types mirror the backend schemas defined in
 * `backend/api/schemas/scheduled_report.py`.
 *
 * @see NEM-3667 - Scheduled Reports Frontend UI
 */

// ============================================================================
// Enums and Literal Types
// ============================================================================

/**
 * Frequency options for scheduled reports.
 */
export type ReportFrequency = 'daily' | 'weekly' | 'monthly';

/**
 * All possible report frequencies.
 */
export const REPORT_FREQUENCIES: readonly ReportFrequency[] = [
  'daily',
  'weekly',
  'monthly',
] as const;

/**
 * Output format options for scheduled reports.
 */
export type ReportFormat = 'pdf' | 'csv' | 'json';

/**
 * All possible report formats.
 */
export const REPORT_FORMATS: readonly ReportFormat[] = ['pdf', 'csv', 'json'] as const;

// ============================================================================
// Scheduled Report Types
// ============================================================================

/**
 * Schema for creating a new scheduled report.
 */
export interface ScheduledReportCreate {
  /** Name/title of the scheduled report */
  name: string;
  /** How often the report should run (daily, weekly, monthly) */
  frequency: ReportFrequency;
  /** Day of week (0=Monday, 6=Sunday) for weekly reports */
  day_of_week?: number | null;
  /** Day of month (1-31) for monthly reports */
  day_of_month?: number | null;
  /** Hour of day to run report (0-23, default 8) */
  hour?: number;
  /** Minute of hour to run report (0-59, default 0) */
  minute?: number;
  /** Timezone for schedule (e.g., 'America/New_York', 'UTC') */
  timezone?: string;
  /** Output format for the report (pdf, csv, json) */
  format?: ReportFormat;
  /** Whether the scheduled report is active */
  enabled?: boolean;
  /** Email addresses to send report to (max 10) */
  email_recipients?: string[] | null;
  /** Include visual charts in the report */
  include_charts?: boolean;
  /** Include detailed event breakdowns */
  include_event_details?: boolean;
}

/**
 * Schema for updating an existing scheduled report.
 * All fields are optional for partial updates.
 */
export interface ScheduledReportUpdate {
  /** Name/title of the scheduled report */
  name?: string;
  /** How often the report should run (daily, weekly, monthly) */
  frequency?: ReportFrequency;
  /** Day of week (0=Monday, 6=Sunday) for weekly reports */
  day_of_week?: number | null;
  /** Day of month (1-31) for monthly reports */
  day_of_month?: number | null;
  /** Hour of day to run report (0-23) */
  hour?: number;
  /** Minute of hour to run report (0-59) */
  minute?: number;
  /** Timezone for schedule (e.g., 'America/New_York', 'UTC') */
  timezone?: string;
  /** Output format for the report (pdf, csv, json) */
  format?: ReportFormat;
  /** Whether the scheduled report is active */
  enabled?: boolean;
  /** Email addresses to send report to (max 10) */
  email_recipients?: string[] | null;
  /** Include visual charts in the report */
  include_charts?: boolean;
  /** Include detailed event breakdowns */
  include_event_details?: boolean;
}

/**
 * Full scheduled report response from API.
 */
export interface ScheduledReport {
  /** Scheduled report ID */
  id: number;
  /** Name/title of the scheduled report */
  name: string;
  /** How often the report runs */
  frequency: ReportFrequency;
  /** Day of week (0=Monday, 6=Sunday) for weekly reports */
  day_of_week: number | null;
  /** Day of month (1-31) for monthly reports */
  day_of_month: number | null;
  /** Hour of day to run report (0-23) */
  hour: number;
  /** Minute of hour to run report (0-59) */
  minute: number;
  /** Timezone for schedule */
  timezone: string;
  /** Output format for the report */
  format: ReportFormat;
  /** Whether the scheduled report is active */
  enabled: boolean;
  /** Email addresses to send report to */
  email_recipients: string[] | null;
  /** Include visual charts in the report */
  include_charts: boolean;
  /** Include detailed event breakdowns */
  include_event_details: boolean;
  /** When the report last ran successfully (ISO 8601) */
  last_run_at: string | null;
  /** When the report is scheduled to run next (ISO 8601) */
  next_run_at: string | null;
  /** When the report was created (ISO 8601) */
  created_at: string;
  /** When the report was last updated (ISO 8601) */
  updated_at: string;
}

/**
 * Response for listing scheduled reports.
 */
export interface ScheduledReportListResponse {
  /** List of scheduled reports */
  items: ScheduledReport[];
  /** Total count */
  total: number;
}

/**
 * Response from manually triggering a report.
 */
export interface ScheduledReportRunResponse {
  /** ID of the report being run */
  report_id: number;
  /** Status of the run (running, queued, failed) */
  status: string;
  /** Status message */
  message: string;
  /** When the run was initiated (ISO 8601) */
  started_at: string;
}

// ============================================================================
// UI Helper Constants
// ============================================================================

/**
 * Human-readable labels for report frequencies.
 */
export const FREQUENCY_LABELS: Record<ReportFrequency, string> = {
  daily: 'Daily',
  weekly: 'Weekly',
  monthly: 'Monthly',
};

/**
 * Human-readable labels for report formats.
 */
export const FORMAT_LABELS: Record<ReportFormat, string> = {
  pdf: 'PDF',
  csv: 'CSV',
  json: 'JSON',
};

/**
 * Day of week labels (0=Monday, 6=Sunday).
 */
export const DAY_OF_WEEK_LABELS: Record<number, string> = {
  0: 'Monday',
  1: 'Tuesday',
  2: 'Wednesday',
  3: 'Thursday',
  4: 'Friday',
  5: 'Saturday',
  6: 'Sunday',
};

/**
 * Common timezone options.
 */
export const COMMON_TIMEZONES: readonly string[] = [
  'UTC',
  'America/New_York',
  'America/Chicago',
  'America/Denver',
  'America/Los_Angeles',
  'Europe/London',
  'Europe/Paris',
  'Europe/Berlin',
  'Asia/Tokyo',
  'Asia/Shanghai',
  'Asia/Singapore',
  'Australia/Sydney',
] as const;

// ============================================================================
// Type Guards
// ============================================================================

/**
 * Type guard to check if a string is a valid report frequency.
 */
export function isReportFrequency(value: unknown): value is ReportFrequency {
  return (
    typeof value === 'string' && REPORT_FREQUENCIES.includes(value as ReportFrequency)
  );
}

/**
 * Type guard to check if a string is a valid report format.
 */
export function isReportFormat(value: unknown): value is ReportFormat {
  return typeof value === 'string' && REPORT_FORMATS.includes(value as ReportFormat);
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Format a time (hour and minute) for display.
 */
export function formatTime(hour: number, minute: number): string {
  const h = hour % 12 || 12;
  const m = minute.toString().padStart(2, '0');
  const ampm = hour < 12 ? 'AM' : 'PM';
  return `${h}:${m} ${ampm}`;
}

/**
 * Get a human-readable schedule description.
 */
export function getScheduleDescription(report: ScheduledReport): string {
  const time = formatTime(report.hour, report.minute);

  switch (report.frequency) {
    case 'daily':
      return `Daily at ${time}`;
    case 'weekly':
      if (report.day_of_week !== null) {
        return `${DAY_OF_WEEK_LABELS[report.day_of_week]}s at ${time}`;
      }
      return `Weekly at ${time}`;
    case 'monthly':
      if (report.day_of_month !== null) {
        const suffix = getOrdinalSuffix(report.day_of_month);
        return `${report.day_of_month}${suffix} of each month at ${time}`;
      }
      return `Monthly at ${time}`;
    default:
      return `At ${time}`;
  }
}

/**
 * Get ordinal suffix for a number (1st, 2nd, 3rd, etc.).
 */
function getOrdinalSuffix(n: number): string {
  const s = ['th', 'st', 'nd', 'rd'];
  const v = n % 100;
  return s[(v - 20) % 10] || s[v] || s[0];
}
